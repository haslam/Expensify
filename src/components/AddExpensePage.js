import React from 'react';
import { connect } from "react-redux";
import ExpenseForm from './ExpenseForm';
import { addExpense } from "../actions/Expenses";

//always use an unconnected component for testing. Hence I'll add an export here
export class AddExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
  }
  onSubmit (expense) {
    //props.dispatch(addExpense(expense)) -- now, we can call onSubmit of mapped to props
    this.props.addExpense(expense);
    this.props.history.push('/');
  }
  render()  {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
}

//For a more testable component, let's use mapDispatchToProps of connect. works similar as mapStateToProps but for Dispatch

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
})

//set first argument [state] to undefined.
export default connect(undefined, mapDispatchToProps)(AddExpensePage)