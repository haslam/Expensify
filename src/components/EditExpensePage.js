import React from 'react';
import { connect } from 'react-redux';
import { matchPath } from 'react-router-dom';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/Expenses';

export class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
    this.onRemove = this.onRemove.bind(this);
  }
  onSubmit (expense) {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
    console.log('Updated', expense) 
  }
  onRemove () {
    this.props.startRemoveExpense({ id: this.props.expense.id} );
    console.log('Removed', this.props.expense.id);
    this.props.history.push('/');
  }
  render() {
  console.log('edit props ', this.props);
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1>Edit Expense</h1>
            <p>Easily make changes to the selected expense.</p> 
          </div>
        </div>
        <div className="content-container">
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
        </div> 
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => {
  const match = matchPath(props.location.pathname, {
    path: '/edit/:id',
    exact: true,
    strict: false
  })
  return expense.id === match.params.id
  })
})
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)

