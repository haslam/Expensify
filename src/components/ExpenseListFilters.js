import React from 'react';
import moment from 'moment';
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/Filters'



export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      calenderFocused: null,
      filterText: '',
      filterSortBy: '',
      startDate: moment().startOf('month'),
      endDate: moment().endOf('month'),
    };
    this.onDatesChange = this.onDatesChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onTextChange = this.onTextChange.bind(this);
    this.onSortChange = this.onSortChange.bind(this);
  }
  onTextChange (e) {
    const sT = e.target.value;
    this.setState(() => ({ filterText: sT}))
    this.props.setTextFilter(sT);
  }
  onSortChange(e) {
    const sortVal = e.target.value;
    this.setState(() => ({ filterSortBy: sortVal}))
    if (sortVal === 'date') {
      this.props.sortByDate();
    }
    else if (sortVal === 'amount') {
      this.props.sortByAmount();
    }
  }
  onDatesChange ({ startDate, endDate }) {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }
  onFocusChange (calenderFocused) {
    this.setState(() => ({ calenderFocused }));
  }

  render() {
   return (
  <div className="content-container">
  <div className="input-group">
    <div className="input-group__item">
      <input type="text" 
        value={this.state.filterText}
        onChange={this.onTextChange}
        className="text-input"
        placeholder="Search expenses"
      />
    </div>
    <div className="input-group__item">
      <select 
        value={this.state.filterSortBy} 
        onChange={this.onSortChange}
        className="select"
      >
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>
    </div>
    <div className="input-group__item">
      <DateRangePicker
        startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
        //startDateId={Math.floor((1 + Math.random() * 0x10000)).toString(16).substring(1)} // PropTypes.string.isRequired,
        startDateId={String(this.props.filters.startDate)}
        endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
        //endDateId={Math.floor((1 + Math.random() * 0x10000)).toString(16).substring(1)} // PropTypes.string.isRequired,
        endDateId={String(this.props.filters.endDate)}
        onDatesChange={this.onDatesChange}
        //   this.props.setStartDate(startDate);
        //   this.props.setEndDate(endDate);
        //   } // PropTypes.func.isRequired,
        // } 
        focusedInput={this.state.calenderFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
        onFocusChange={calenderFocused => this.setState({ calenderFocused })} // PropTypes.func.isRequired,
        numberOfMonths={1}
        showDefaultInputIcon={true}
        showClearDates={true}
      />
    </div>
  </div>
  </div>
    );
  }
}

const mapStateToProps = (state) => ({
    filters: state.filters
})

const mapDispatchToProps = (dispatch, props) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate())
})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

{/* <DateRangePicker 
startDate={this.props.filters.startDate}
endDate={this.props.filters.endDate}
focusedInput={this.state.calenderFocused}
onFocusChange={this.onFocusChange}
showClearDates={true}
numberOfMonths={1}
isOutsideRange={() => false}      
/> */}