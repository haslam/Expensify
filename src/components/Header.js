import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { beginLogout } from '../actions/auth'
import hashEmail from '../helper/hashing'

export const Header = ({ beginLogout, email, displayName }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <div className="show-for-mobile">
          <button className="mobile__hamburger" role="menu" title="Navigation menu hamburger" onClick={openNavigation}>{`\u{1F354}`}</button>
        </div>
        <Link className="header__title" to="/dashboard">
          <h1>Expensified</h1>
        </Link>
        
        <div className="header__profile show-for-desktop">
          <div className="header__avatar">
            <img className="avatar avatar--sphere" src={`https://gravatar.com/avatar/${hashEmail(email)}?s=200`}/>
            <small style={{paddingRight: "1rem", color: "#dcdcdc"}}>{displayName}</small>
          </div>
          <div style={{ alignSelf: "center" }}>
            <button className="button button--trans" onClick={beginLogout}>Logout</button>
          </div>  
        </div>
      </div>
    </div>
  </header>
)

const openNavigation = () => {
  document.body.classList.toggle('mobile__nav--on');
}

const mapDispatchToProps = (dispatch) => ({
  beginLogout: () => dispatch(beginLogout())
})

const mapStateToProps = (state) => ({
  email: state.auth.email,
  displayName: state.auth.displayName,
})
export default connect(mapStateToProps, mapDispatchToProps)(Header)

{/* <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink> */}
