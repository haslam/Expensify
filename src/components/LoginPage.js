import React from "react";
import { connect } from "react-redux";
import { beginLogin } from "../actions/auth";
export const LoginPage = ({ beginLogin }) => (
  <div>
    <h2>Login with your username and password</h2>
      <button onClick={ beginLogin }>Login</button>
  </div>
)


const mapDispatchToProps = (dispatch) => ({
  beginLogin: () => dispatch(beginLogin())
})

export default connect(undefined, mapDispatchToProps)(LoginPage)

{/* <input type="text" name="username" placeholder="username"/>
<input type="password" name="password" placeholder="******" /> */}