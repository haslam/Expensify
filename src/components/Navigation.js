import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTachometerAlt, faMoneyCheckAlt, faPoll, faChartLine } from "@fortawesome/free-solid-svg-icons";
import { beginLogout } from '../actions/auth'
import hashEmail from '../helper/hashing'

export const Navigation = ({ beginLogout, email, displayName}) => {

  return (
    <div className="mobile-navigation mobile-navigation--left">
      <div className="navigation__profile">
        <img className="avatar avatar--sphere" src={`https://gravatar.com/avatar/${hashEmail(email)}?s=200`}/>
        <small>{displayName}</small>
        <button className="button__logout--left" onClick={beginLogout}>Logout</button>
      </div>
      <div className="navigation__tabs">
        <ul role="navigation" className="tabs-menu">
          <li className="tabs-menu__list">
            <Link to="/dashboard" activeClassName="tm-active">
              <FontAwesomeIcon icon={faTachometerAlt} size="2x" />
            </Link>
          </li>
          <li className="tabs-menu__list">
            <Link to="/expenses" activeClassName="tm-active">
              <FontAwesomeIcon icon={faMoneyCheckAlt} size="2x" />
            </Link>
          </li>
          <li className="tabs-menu__list">
            <Link to="/report" activeClassName="tm-active">
              <FontAwesomeIcon icon={faPoll} size="2x" />
            </Link>
          </li>
        </ul>
      </div>
    </div>
  )
}

const mapDispatchToProps = (dispatch) => ({
  beginLogout: () => dispatch(beginLogout())
})
const mapStateToProps = (state) => ({
  email: state.auth.email,
  displayName: state.auth.displayName
})

export default connect(mapStateToProps, mapDispatchToProps)(Navigation)