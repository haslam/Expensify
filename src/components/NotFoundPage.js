import React from 'react';

export default () => (
  <p>
    You've hit a wrong path
  </p>
)