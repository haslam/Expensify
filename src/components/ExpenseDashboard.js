import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from "./ExpensesSummaries";
import AddExpense from './AddExpensePage'
import Header from './Header';

const ExpenseDashboardPage = () => (
  <div>
    <Header />
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)

export default ExpenseDashboardPage
