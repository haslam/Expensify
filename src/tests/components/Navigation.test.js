import React from 'react'
import { shallow } from 'enzyme'
import { Navigation } from '../../components/Navigation';
import hashEmail from '../../helper/hashing';

describe('Side navigation ', () => {
  let wrapper, logout;
  const email = 'test@example.com';

  it('should render correctly', () => {
    wrapper = shallow(<Navigation email={email} beginLogout={() => {}} />);
    expect(wrapper).toMatchSnapshot();
  })

  it('should call beginLogout when button is clicked', () => {
    logout = jest.fn();
    wrapper = shallow(<Navigation email={email} beginLogout={logout} />);
    wrapper.find('.button__logout--left').simulate('click');
    expect(logout).toHaveBeenCalled();
  })
})