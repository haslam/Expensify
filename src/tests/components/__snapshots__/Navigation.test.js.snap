// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Side navigation  should render correctly 1`] = `
<div
  className="mobile-navigation mobile-navigation--left"
>
  <div
    className="navigation__profile"
  >
    <img
      className="avatar avatar--sphere"
      src="https://gravatar.com/avatar/55502f40dc8b7c769880b10874abc9d0?s=200"
    />
    <small />
    <button
      className="button__logout--left"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="navigation__tabs"
  >
    <ul
      className="tabs-menu"
      role="navigation"
    >
      <li
        className="tabs-menu__list"
      >
        <Link
          activeClassName="tm-active"
          replace={false}
          to="/dashboard"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  576,
                  512,
                  Array [],
                  "f3fd",
                  "M288 32C128.94 32 0 160.94 0 320c0 52.8 14.25 102.26 39.06 144.8 5.61 9.62 16.3 15.2 27.44 15.2h443c11.14 0 21.83-5.58 27.44-15.2C561.75 422.26 576 372.8 576 320c0-159.06-128.94-288-288-288zm0 64c14.71 0 26.58 10.13 30.32 23.65-1.11 2.26-2.64 4.23-3.45 6.67l-9.22 27.67c-5.13 3.49-10.97 6.01-17.64 6.01-17.67 0-32-14.33-32-32S270.33 96 288 96zM96 384c-17.67 0-32-14.33-32-32s14.33-32 32-32 32 14.33 32 32-14.33 32-32 32zm48-160c-17.67 0-32-14.33-32-32s14.33-32 32-32 32 14.33 32 32-14.33 32-32 32zm246.77-72.41l-61.33 184C343.13 347.33 352 364.54 352 384c0 11.72-3.38 22.55-8.88 32H232.88c-5.5-9.45-8.88-20.28-8.88-32 0-33.94 26.5-61.43 59.9-63.59l61.34-184.01c4.17-12.56 17.73-19.45 30.36-15.17 12.57 4.19 19.35 17.79 15.17 30.36zm14.66 57.2l15.52-46.55c3.47-1.29 7.13-2.23 11.05-2.23 17.67 0 32 14.33 32 32s-14.33 32-32 32c-11.38-.01-20.89-6.28-26.57-15.22zM480 384c-17.67 0-32-14.33-32-32s14.33-32 32-32 32 14.33 32 32-14.33 32-32 32z",
                ],
                "iconName": "tachometer-alt",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size="2x"
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Link>
      </li>
      <li
        className="tabs-menu__list"
      >
        <Link
          activeClassName="tm-active"
          replace={false}
          to="/expenses"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  640,
                  512,
                  Array [],
                  "f53d",
                  "M608 32H32C14.33 32 0 46.33 0 64v384c0 17.67 14.33 32 32 32h576c17.67 0 32-14.33 32-32V64c0-17.67-14.33-32-32-32zM176 327.88V344c0 4.42-3.58 8-8 8h-16c-4.42 0-8-3.58-8-8v-16.29c-11.29-.58-22.27-4.52-31.37-11.35-3.9-2.93-4.1-8.77-.57-12.14l11.75-11.21c2.77-2.64 6.89-2.76 10.13-.73 3.87 2.42 8.26 3.72 12.82 3.72h28.11c6.5 0 11.8-5.92 11.8-13.19 0-5.95-3.61-11.19-8.77-12.73l-45-13.5c-18.59-5.58-31.58-23.42-31.58-43.39 0-24.52 19.05-44.44 42.67-45.07V152c0-4.42 3.58-8 8-8h16c4.42 0 8 3.58 8 8v16.29c11.29.58 22.27 4.51 31.37 11.35 3.9 2.93 4.1 8.77.57 12.14l-11.75 11.21c-2.77 2.64-6.89 2.76-10.13.73-3.87-2.43-8.26-3.72-12.82-3.72h-28.11c-6.5 0-11.8 5.92-11.8 13.19 0 5.95 3.61 11.19 8.77 12.73l45 13.5c18.59 5.58 31.58 23.42 31.58 43.39 0 24.53-19.05 44.44-42.67 45.07zM416 312c0 4.42-3.58 8-8 8H296c-4.42 0-8-3.58-8-8v-16c0-4.42 3.58-8 8-8h112c4.42 0 8 3.58 8 8v16zm160 0c0 4.42-3.58 8-8 8h-80c-4.42 0-8-3.58-8-8v-16c0-4.42 3.58-8 8-8h80c4.42 0 8 3.58 8 8v16zm0-96c0 4.42-3.58 8-8 8H296c-4.42 0-8-3.58-8-8v-16c0-4.42 3.58-8 8-8h272c4.42 0 8 3.58 8 8v16z",
                ],
                "iconName": "money-check-alt",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size="2x"
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Link>
      </li>
      <li
        className="tabs-menu__list"
      >
        <Link
          activeClassName="tm-active"
          replace={false}
          to="/report"
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  448,
                  512,
                  Array [],
                  "f681",
                  "M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zM160 368c0 8.84-7.16 16-16 16h-32c-8.84 0-16-7.16-16-16V240c0-8.84 7.16-16 16-16h32c8.84 0 16 7.16 16 16v128zm96 0c0 8.84-7.16 16-16 16h-32c-8.84 0-16-7.16-16-16V144c0-8.84 7.16-16 16-16h32c8.84 0 16 7.16 16 16v224zm96 0c0 8.84-7.16 16-16 16h-32c-8.84 0-16-7.16-16-16v-64c0-8.84 7.16-16 16-16h32c8.84 0 16 7.16 16 16v64z",
                ],
                "iconName": "poll",
                "prefix": "fas",
              }
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size="2x"
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </Link>
      </li>
    </ul>
  </div>
</div>
`;
